babel/messages/extract.py

```diff
@@ -516,6 +511,29 @@ def extract_python(fileobj, keywords, comment_tags, options):
             funcname = value
 
 
+def _parse_python_string(value, encoding, future_flags):
+    # Unwrap quotes in a safe manner, maintaining the string's encoding
+    # https://sourceforge.net/tracker/?func=detail&atid=355470&aid=617979&group_id=5470
+    code = compile(
+        f'# coding={str(encoding)}\n{value}',
+        '<string>',
+        'eval',
+        ast.PyCF_ONLY_AST | future_flags,
+    )
+    if not isinstance(code, ast.Expression):
+        return None
+    body = code.body
+    if isinstance(body, ast.Str):
+        return body.s
+    if isinstance(body, ast.JoinedStr):  # f-string
+        if all(isinstance(node, ast.Str) for node in body.values):
+            return ''.join(node.s for node in body.values)
+        if all(isinstance(node, ast.Constant) for node in body.values):
+            return ''.join(str(node.value) for node in body.values)
```

Comment:
From what I can tell from trying things out, it seems like the AST always combines all the strings into a single constant. If that's the case maybe this could be simplified?

```py
>>> value='"123" f"456" "789"'
>>> ast.dump(compile(f'# coding={str(encoding)}\n{value}', '<string>', 'eval', ast.PyCF_ONLY_AST))
"Expression(body=JoinedStr(values=[Constant(value='123456789')]))"
```

I'm not too familiar with all the use cases of this function, but from looking at the examples and docstrings, it seems like it should work.

Reply:
Yeah, that's the case, but I suspect that's just an optimization in the parser that we shouldn't necessarily rely on. (Sure, it's not likely that it will be un-optimized anytime soon, but perhaps if Babel is used with a Python implementation that doesn't do that..?)